### ISSUE TITLE
A FunctionExpression/Method/Getter/Setter may capture its own name

### ISSUE THREAD

**Issue by Yoric** [binast/ecmascript-binary-ast#23]
The following snippet demonstrate how the name of a `FunctionExpression` can be captured:

```js
let a = function f() {
  return function g() {
    return f;
  }
};
let b = a()(); // This is `f`
```

The same technique would probably work for `Method`, `Getter`, `Setter`.

For the moment, we have nowhere to store this information.

**Comment by syg** [comment]
I think this is only a bug for named `FunctionExpression`s. `FunctionDeclaration`s' names are treated as bindings in the enclosing scope. `Method`, `Getter`, and `Setter` can't be recursively called like that, since they don't make a binding.

**Comment by Yoric** [comment]
So, if I read this correctly, we store the information in the `parameterScope`, right? This would probably deserve a comment in the webidl section.

**Comment by syg** [comment]
> So, if I read this correctly, we store the information in the parameterScope, right? This would probably deserve a comment in the webidl section.

That's right, and it's mainly because of how the declared names are checked. AssertedParameterScopes' declared names are checked with the BoundNames static semantics in ecma262, instead of VarDeclaredNames or LexicallyDeclaredNames. That happens to match what the Catch production's runtime semantics already does.
